### PR TITLE
Improvement/better game scoreboard

### PR DESCRIPTION
**Pull request by RedstoneFuture**
**This pull-request add an improved scoreboard management system.**

- make scoreboard configuration (`config.yml`) user-friendly with more options
```
sidebar:
  title: '&eInfo ●&6•'
  member_list_style: '%team_color%%playername%'
  member_list_max: '4'
  entries:
  - '§7Time left:'
  - §e» %time%m
  - ''
  - '%team1% §7» %team1_color%%team1_amount%'
  - ''
  - '%team2% §7» %team2_color%%team2_amount%'
```

- more Placeholders
```
%team1%, %team2%
%team1_color%, %team2_color%
%team1_amount%, %team2_amount%
%team1_members%, %team2_members%
%arena_name%
%time%
```

- nice look: dynamic member list (`%team1_members%` and `%team2_members%`), always with only the required number of lines.
grafik

- usefully scoreboard updates in the lobby and arena, so it's always up-to-date

- tested on my production server and works fine

### CODE REVIEW

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/wrapper/game/Team.java [hunk 1]
```diff
@@ -78,11 +78,8 @@ public boolean removeMember(MWPlayer player) {
         Player p = player.getPlayer();
         player.setTeam(null);
         if (p != null) {
-            if (scoreboardTeam.hasPlayer(p))
-                scoreboardTeam.removePlayer(p);
-
-            game.getScoreboard().getTeam("2Guest§7").addPlayer(p);
             p.setDisplayName("§7" + p.getName() + "§r");
+            // TODO remove scoreboard
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/wrapper/game/Team.java]: Could you explain what you mean by this comment? And what is there still to do?

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java [hunk 2]
```diff
@@ -149,13 +155,9 @@ private static void saveConfig() {
...
 
-    public static HashMap<String, Integer> getScoreboardEntries() {
-        HashMap<String, Integer> ret = new HashMap<>();
-        ConfigurationSection section = cfg.getConfigurationSection("sidebar.entries");
-        for (String s : section.getKeys(false)) {
-            ret.put(section.getString(s), Integer.valueOf(s));
-        }
-        return ret;
+    public static List<String> getScoreboardEntries() {
+        List<String> scoreboardLines = cfg.getStringList("sidebar.entries");
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java]: Please inline this variable (`return cfg.getStringList("...");`)

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java [hunk 3]
```diff
@@ -195,103 +197,90 @@ public static YamlConfiguration getConfig() {
...
-    public static String getArenaFolder() {
-        return cfg.getString("arena_folder") + "/";
-    }
-
-    public static boolean isContactAuth() {
-        return cfg.getBoolean("contact_auth_server");
-    }
+    public static String getScoreboardTitle() {return cfg.getString("sidebar.title");}
+    public static String getScoreboardMembersStyle() {return cfg.getString("sidebar.member_list_style");}
+    public static int getScoreboardMembersMax() {return cfg.getInt("sidebar.member_list_max");}
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java]: Please stick to the java formatting and add lines inbetween methods

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java [hunk 4]
```diff
@@ -56,10 +55,8 @@ public void onJoin(PlayerArenaJoinEvent e) {
         Bukkit.getScheduler().runTaskLater(MissileWars.getInstance(), () -> p.teleport(game.getArena().getSpectatorSpawn()), 2);
 
         Bukkit.getScheduler().runTaskLater(MissileWars.getInstance(), () -> p.setGameMode(GameMode.SPECTATOR), 35);
-        Scoreboard sb = game.getScoreboard();
-        p.setScoreboard(sb);
-        sb.getTeam("2Guest§7").addPlayer(p);
         p.setDisplayName("§7" + p.getName() + "§r");
+        // TODO remove scoreboard
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java]: Same as in Team.java, why is TODO comment?

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 5]
```diff
@@ -20,57 +20,254 @@
...
-    private final Scoreboard board;
 
-    public void updateInGameScoreboard() {
-        removeScoreboard();
+    private Team team1;
+    private Team team2;
+    private Arena arena;
+
+    // get config options
+    private String scoreBoardTitle = Config.getScoreboardTitle();
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: Please make final what can be final

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 6]
```diff
@@ -20,57 +20,254 @@
...
+    private String memberListStyle = Config.getScoreboardMembersStyle();
+    private int memberListMaxSize = Config.getScoreboardMembersMax();
+    private List<String> scoreBoardEntries = Config.getScoreboardEntries();
+
+    private boolean isTeam1ListUsed = false;
+    private boolean isTeam2ListUsed = false;
+
+    private int maxScoreboardLines = 15;
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: Please declare constants as private, final and static and name it UPPERCASE. Also, why is this not used?

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 7]
```diff
@@ -20,57 +20,254 @@
...
+        if (teams.size() < line) {
+            team = board.registerNewTeam(arena.getName() + "-" + line);
+            team.addEntry("§" + colorCodes[line - 1]);
+            obj.getScore("§" + colorCodes[line - 1]).setScore(line);
+            teams.put(line, team);
 
-        for (String entry : entries.keySet()) {
-            String s = rep(entry);
-            obj.getScore(s).setScore(entries.get(entry));
+            // VersionUtil.setScoreboardTeamColor(t, ChatColor.getByChar(team2.getColorCode().charAt(1)));
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: Why did you comment out this line? Please add either an explanation or remove it entirely

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 8]
```diff
@@ -20,57 +20,254 @@
...
+     *
+     * @return (int) the amount of offset lines
+     */
+    private int getLineOffset() {
+
+        int team1ListSize = 0;
+        int team2ListSize = 0;
+
+        if (isTeam1ListUsed) {
+            if (team1.getMembers().size() > memberListMaxSize) {
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: Please replace if/else with `team1ListSize = Math.min(team1.getMembers().size(), memberListMaxSize);`

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 9]
```diff
@@ -20,57 +20,254 @@
...
+                team1ListSize = memberListMaxSize;
+            } else {
+                team1ListSize = team1.getMembers().size();
+            }
+            team1ListSize--;
+        }
+
+        if (isTeam2ListUsed) {
+            if (team2.getMembers().size() > memberListMaxSize) {
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: Same as above

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 10]
```diff
@@ -20,57 +20,254 @@
...
+            } else if (cleanLine.contains("%team2_members%")) {
+                isTeam2ListUsed = true;
+            }
+        }
+
+        updateScoreboard();
+    }
+
+    /**
+     * This method create a team for the scoreboard and add it to the teams ArrayList.
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: create -> creates, add -> adds

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 11]
```diff
@@ -20,57 +20,254 @@
...
+                scoreboardLine--;
+            }
+        }
     }
 
-    private String rep(String entry) {
-        return replaceTeam1(replaceTeam2(replaceTime(entry)));
+    /**
+     * This method calculates the offset lines based of the amount of players
+     * and the using of the member-list placeholders for booth teams.
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: booth -> both

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 12]
```diff
@@ -20,57 +20,254 @@
...
+        if (isTeam2ListUsed) {
+            if (team2.getMembers().size() > memberListMaxSize) {
+                team2ListSize = memberListMaxSize;
+            } else {
+                team2ListSize = team2.getMembers().size();
+            }
+            team2ListSize--;
+        }
+
+        return (team1ListSize + team2ListSize);
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: (team1ListSize + team2ListSize) -> team1ListSize + team2ListSize

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 13]
```diff
@@ -20,57 +20,254 @@
...
+        int scoreboardLine = scoreBoardEntries.size() + getLineOffset();
+
+        // add new teams
+        for (int i = 1; i <= scoreboardLine; i++) {
+            addScoreboardTeam(i);
+        }
+
+        /*
+
+        Not possible! When a user leaves the game, the scoreboard must be reset.
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: What is this code doing here?

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 14]
```diff
@@ -20,57 +20,254 @@
...
+                } else {
+                    placeholderTeam = team2;
+                }
+
+                // check if there is no one in the team at the moment
+                if (placeholderTeam.getMembers().isEmpty()) {
+                    continue;
+                }
+
+                int players = 0;
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: maybe rename variable to `currentPlayerIndex` or something

**Review by daniel-naegele — CHANGES_REQUESTED** [review]

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/wrapper/game/Team.java (hunk 1 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/wrapper/game/Team.java:82]: It's a deprecated method.

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java (hunk 4 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java:59]: It's a deprecated method.

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java (hunk 3 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java]: It is currently grouped by topic.

grafik

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java [hunk 15]
```diff
@@ -68,21 +68,19 @@ public void onRespawn(PlayerRespawnEvent e) {
...
     @EventHandler
-
     public void onDeath(PlayerDeathEvent e) {
-        if (!isInLobbyArea(e.getEntity().getLocation()))
-            return;
+        if (!isInLobbyArea(e.getEntity().getLocation())) return;
 
         Player p = e.getEntity();
-        p.setHealth(p.getMaxHealth());
+        p.setHealth(p.getAttribute(Attribute.GENERIC_MAX_HEALTH).getDefaultValue());
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/EndListener.java]: I think that this is not compatible with 1.8 and some of the higher versions. Please stick to `p.setHealth(p.getMaxHealth());`

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java [hunk 16]
```diff
@@ -279,7 +266,7 @@ public void onDeath(PlayerDeathEvent e) {
 
         // spectator respawn for people there are not in a team
         if (player.getTeam() == null) {
-            p.setHealth(p.getMaxHealth());
+            p.setHealth(p.getAttribute(Attribute.GENERIC_MAX_HEALTH).getDefaultValue());
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/GameListener.java]: Same as in EndListener

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 17]
```diff
@@ -186,7 +174,7 @@ public void updateScoreboard() {
 
     /**
      * This method calculates the offset lines based of the amount of players
-     * and the using of the member-list placeholders for booth teams.
+     * and the using of the member-list placeholders for booths teams.
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: As in the old review: booth (booths) -> both

**Review by daniel-naegele — CHANGES_REQUESTED** [review]

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java (hunk 3 above)
Comment (daniel-naegele) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/Config.java]: The grouping is not the problem, but rather the formatting, just hit Strg + Alt + L once in this file (and overall in this project)

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java [hunk 18]
```diff
@@ -20,57 +20,242 @@
...
+    private boolean isTeam2ListUsed = false;
+
+    private int maxScoreboardLines = 15;
+    public Scoreboard board;
+    private Objective obj;
+    private HashMap<Integer, org.bukkit.scoreboard.Team> teams = new HashMap<>();
+    String[] colorCodes = {"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f"};
+    
+    /**
+     * This method register the scoreboard.
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/util/ScoreboardManager.java]: register -> registers (He, she, it, dass "S" muss mit)

**Review by daniel-naegele — APPROVED** [review]

**Pull request merged**